include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,383 @@
+
+            quant_type D = ndf.template D<sample_type, Interaction, MicrofacetCache>(qq, _sample, interaction, cache);
+            scalar_type DG = D.projectedLightMeasure;
+            if (D.microfacetMeasure < bit_cast<scalar_type>(numeric_limits<scalar_type>::infinity))
+            {
+                g2g1_query_type gq = ndf.template createG2G1Query<sample_type, Interaction>(_sample, interaction);
+                DG *= ndf.template correlated<sample_type, Interaction>(gq, _sample, interaction);
+            }
+            NBL_IF_CONSTEXPR(IsBSDF)
+                return impl::__implicit_promote<spectral_type, typename fresnel_type::vector_type>::__call(_f(hlsl::abs(cache.getVdotH()))) * DG;
+            else
+                return impl::__implicit_promote<spectral_type, typename fresnel_type::vector_type>::__call(_f(cache.getVdotH())) * DG;
+        }
+        else
+            return hlsl::promote<spectral_type>(0.0);
+    }
+    template<typename C=bool_constant<!IsAnisotropic> >
+    enable_if_t<C::value && !IsAnisotropic, spectral_type> eval(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
+    {
+        return __eval<isotropic_interaction_type, isocache_type>(_sample, interaction, cache);
+    }
+    spectral_type eval(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(anisocache_type) cache)
+    {
+        return __eval<anisotropic_interaction_type, anisocache_type>(_sample, interaction, cache);
+    }
+
+    template<typename C=bool_constant<!IsBSDF> >
+    enable_if_t<C::value && !IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, const vector2_type u, NBL_REF_ARG(anisocache_type) cache)
+    {
+        ray_dir_info_type localV_raydir = interaction.getV().transform(interaction.getToTangentSpace());
+        const vector3_type localV = interaction.getTangentSpaceV();
+        const vector3_type localH = ndf.generateH(localV, u);
+
+        cache = anisocache_type::createForReflection(localV, localH);
+        struct reflect_wrapper
+        {
+            vector3_type operator()() NBL_CONST_MEMBER_FUNC
+            {
+                return r(VdotH);
+            }
+            bxdf::Reflect<scalar_type> r;
+            scalar_type VdotH;
+        };
+        reflect_wrapper r;
+        r.r = bxdf::Reflect<scalar_type>::create(localV, localH);
+        r.VdotH = cache.getVdotH();
+        ray_dir_info_type localL = localV_raydir.template reflect<reflect_wrapper>(r);
+
+        // fail if samples have invalid paths
+        if (localL.getDirection().z < scalar_type(0.0))  // NdotL<0
+            localL.makeInvalid(); // should check if sample direction is invalid
+
+        return sample_type::createFromTangentSpace(localL, interaction.getFromTangentSpace());
+    }
```

Comment:
btw you're not checking at all that the `NdotV>minimum` which will lead you to call the generateH function with samples below the hemisphere

this in turn for a rough NDF will result in some `localL.getDirection()` in the upper hemisphere (reflection off a microfacet in the upward direction) and not failing